# TP : calcul matriciel

On donne ci-dessous deux variables A1 et A2, représentant deux matrices

In [1]:
A1=[[1,2,3],
    [4,5,6],
    [7,8,9]]

A2=[[1,4,7],
    [2,5,8],
    [3,6,9]]

##### question 1

Définir deux variables L et C pour définir les matrices suivantes : 

$$L = \begin{pmatrix}
    5 & 9 & -2
  \end{pmatrix}  $$

$$ C = \begin{pmatrix}
    -1\\
    4\\
    2
  \end{pmatrix}
$$

In [3]:
L = [ [ 5,9,-2 ] ]

C = [ [-1], [4], [2] ]

## Somme de deux matrices


##### question 2 
Compléter le code suivant pour qu'il puisse calculer la somme de deux matrices de même format.

In [7]:
def somme(M,N):
    nblignesM = len(M)
    nbcolM = len(M[0])
    nblignesN = len(N)
    nbcolN = len(N[0])
    
    #if  not( nblignesM == nblignesN  and  nbcolM == nbcolN  ):
    if ( nblignesM != nblignesN  or  nbcolM != nbcolN  ):
        print("Matrices de dimensions differentes, calcul impossible !")
    else : 
        resu =[[0 for j in range(nbcolM)] for i in range(nblignesM)]
        for i in range(nblignesM):
            for j in range(nbcolM):
                resu[i][j]= M[i][j] + N[i][j]  # A COMPLETER
        return resu        

Tester la fonction sur les matrices définies précédemment

In [8]:
somme(A1,A2)

[[2, 6, 10], [6, 10, 14], [10, 14, 18]]

In [9]:
somme(C,C)

[[-2], [8], [4]]

In [10]:
somme(L,C)

Matrices de dimensions differentes, calcul impossible !


## Produit de deux matrices

##### question 3
Compléter le code suivant pour qu'il puisse calculer le produit de deux matrices de formats compatibles.

In [11]:
def produit_matrices(M,N):
# tous les ... sont à compléter
    n=len(M)    # nombre de lignes de M
    p=len(M[0]) # nombre de colonnes de M
    q=len(N[0]) # nombre de colonnes de N
    if (len(N) == p):
        resu=[ [ 0 for j in range(q)] for i in range(n)]
        for i in range(n):
            for j in range(q):
                for k in range(p) : 
                    resu[i][j] += M[i][k] * N[k][j]
        return resu
    else : 
        print("Erreur de format")

Tester la fonction sur les matrices définies précédemment

In [13]:
produit_matrices(A1,C)

[[13], [28], [43]]

In [14]:
produit_matrices(A1,A2)

[[14, 32, 50], [32, 77, 122], [50, 122, 194]]

##### question 4
En utilisant les questions précédents, calculer : 
1. A1 * A2
2. L * C
3. C * L + A2 * A1

In [15]:
produit_matrices(A1,A2)

[[14, 32, 50], [32, 77, 122], [50, 122, 194]]

In [17]:
produit_matrices(L,C)

[[27]]

In [18]:
somme(produit_matrices(C,L) , produit_matrices(A2,A1))

[[61, 69, 92], [98, 129, 100], [100, 126, 122]]

#### Utilisation
La fabrication d'un pain viennois nécessite 100g de farine, 20 g de beurre et 10g de sucre. La fabrication d'un croissant nécessite 30g de farine, 12 g de beurre et 5g de sucre.  La fabrication d'une galette sèche nécessite 250g de farine, 125 g de beurre et 30g de sucre. 

Un client commande 12 pains, 50 croissants, et 10 galettes à livrer Lundi.
Il commande aussi 15 pains, 40 croissants et 20 galettes à livrer Mardi.


|  | pain viennois | croissant | galette |
:-|:-------------:|:---------:|:-------:|
|farine |  100 |30 |250 
|beurre |20|12|125
|sucre | 10|5| 30



|     |Lundi|Mardi|
:-----|:---:|:---:|
|pain |12   |15   |
|croissant|50|40|  
|galette|10|20|


##### question 5.a
Saisir les tableaux de données dans deux variables en python

In [19]:
ingredients = [[100, 30, 250 ],
               [20 , 12, 125 ],
               [10 , 5 , 30  ]]
jours = [[12 ,15 ],
         [50 ,40 ],
         [10 ,20 ]]

##### question 5.b
Calculer et afficher le produit des deux matrices saisies dans la question précédente

In [20]:
produit_matrices(ingredients , jours)

[[5200, 7700], [2090, 3280], [670, 950]]

##### question 5.c
Expliquer ce que représentent les valeurs trouvées .

|     |Lundi|Mardi|
:-----|:---:|:---:|
|farine |5200 |7700 |
|beurre|2090|3280|  
|sucre|670|950|